packet_link_qualification/index.md

```diff
@@ -0,0 +1,734 @@
+// the exact error code and message.
+// Expected errors codes:
+// NOT_FOUND when the requested id was not found by the service.
+// INVALID_ARGUMENT for any configuration parameter which is unsupported.
+message GetResponse{
+  map<string, QualificationResult> results = 1;
+}
+
+// States of qualification.
+enum QualificationState {
+  QUALIFICATION_STATE_UNSPECIFIED = 0;
+  QUALIFICATION_STATE_ERROR = 1;  // The qualification has errored.
+  QUALIFICATION_STATE_IDLE = 2;       // Initial state for the qualification.
+  QUALIFICATION_STATE_SETUP = 3;       // Interface is being configured.
+  QUALIFICATION_STATE_RUNNING = 4;    // Qualification underway.
+  QUALIFICATION_STATE_TEARDOWN = 5;    // Interface is being reset.
+  QUALIFICATION_STATE_COMPLETED = 6;  // Qualification is complete.
+}
+
+message QualificationResult {
+  // ID of the qualification.
+  string id = 1;
+
+  // Interface name of the qualification.
+  string interface_name = 2;
+
+  // The state the test was in when the results were snapshotted.
+  QualificationState state = 3;
+
+  // The number of qualification packets sent.
+  uint64 packets_sent = 4;
+
+  // The number of qualification packets received.
+  uint64 packets_received = 5;
+
+  // The number of packets transmitted that experienced corruption.
+  uint64 packets_error = 6;
+
+  // The number of packets dropped by the device due to internal drop,
+  // lookup or forwarding errors.
+  uint64 packets_dropped = 7;
+
+  // The qualification start time. Also when the first snapshot of
+  // results are taken.
+  google.protobuf.Timestamp start_time = 8;
+
+  // The qualification end time or the current snapshot time since epoch.
+  google.protobuf.Timestamp end_time = 9;
+
+  // Expected rate for the qualification. This is the computed or
+  // observed rate that the service expected to be maintained
+  // throughout the qualification duration.
+  uint64 expected_rate_bytes_per_second = 10;
+
+  // The qualification rate observed during the qualification. 
+  // It is updated every update_interval in bytes per second.
+  uint64 qualification_rate_bytes_per_second = 11;
+
+  // Status response for the Qualification result. Will only be set if the
+  // state is QUALIFICATION_STATE_ERROR.
+  google.rpc.Status status = 12;
+}
+```
+
+#### Delete Messages
+
+```
+// DeleteRequest will delete the qualification results for the provided id.
+message DeleteRequest{
+  repeated string ids = 1;
+}
+
+// Delete response will contain a list of all id's in the request to be deleted.
+// If the id was not found NOT_FOUND will be returned.
+message DeleteResponse{
+  map<string, google.rpc.Status> results = 1;
+}
+
+```
+
+#### Capabilities Messages
+
+```
+message CapabilitiesRequest {
+}
+
+message CapabilitiesResponse {
+  // Current timestamp on the service.
+  google.protobuf.Timestamp time = 1;
+  // Does the service support NTP based synchronization.
+  bool ntp_synced = 2;
```

Comment:
Is the semantic here "I support NTP sync", or is it "I am /currently/ NTP sync'd"? I think the latter seems more useful since support != actually configured.

Reply:
I guess the `time` field covers some of this -- do we need both? If the device reports the correct time now, does it matter where it got the sync from? The only case that I can think of that it would is where we are scheduling a qualification a long way into the future and drift /might/ happen, which seems a corner case.